iwyu_output.cc

```diff
@@ -1844,6 +1881,29 @@ size_t IwyuFileInfo::CalculateAndReportIwyuViolations() {
   // we *do* need to add it.
   set<string> associated_desired_includes = AssociatedDesiredIncludes();
 
+  // Resolve using declarations before continuing.  This handles the case
+  // where there's a using declaration in the file but no code is actually
+  // using it. If that happens, we might try to remove all of the headers with
+  // the decls that the using decl is referencing, which would result in a
+  // compilation error at best. A possible solution is to remove the using
+  // decl if it's not used, but that doesn't work for header files because a
+  // using decl in a header is an exported symbol, so we don't want to do that
+  // either. As a compromise, we arbitrarily add the first shadow decl to make
+  // sure everything still compiles instead of removing the using decl. A
+  // more thorough approach would be to scan the current list of includes that
+  // alredy name this decl (like in the overloaded function case) and include
+  // one of those so we don't include a file we don't actually need.
+  for (typename map<const UsingDecl*, bool>::value_type using_decl_status
+       : using_decl_referenced_) {
+    if (!using_decl_status.second) {
+      const UsingDecl* using_decl = using_decl_status.first;
+      ReportForwardDeclareUse(using_decl->getUsingLoc(),
+                              using_decl->shadow_begin()->getTargetDecl(), 
+                              /* in_cxx_method_body */ false,
+                              "(for un-referenced using)");
+    }
+  }
```

Comment:
> Does it matter that you are doing this after AssociatedDesiredIncludes, not before?

Curious, I would think it would matter.  I'll write a test to see if I can break it.

> And it feels a little bit dirty doing it in CalculateAndReportIwyuViolations. Maybe we should add something like IwyuFileInfo::ResolvePendingAnalysis or IwyuFileInfo::DidFinishFileProcessing and call it from HandleTranslationUnit. What do you think?

I agree.  I like ResolvePendingAnalysis.  I'll make that change.